### DISCUSSION TITLE
Version 2.0

### DISCUSSION THREAD

**Discussion by oskvr37** [oskvr37/tiddl#51] [Announcements]
I'm developing version 2.0, which will feature:

- Enhanced CLI experience
- Advanced Tidal search functionality
- Access to your Tidal favorites, including tracks, albums, artists, and playlists

**Comment by Darkloke** [comment]
So... 2.0. is here i suppose. What are the new changes? I tried to find some, but saw only this - am i missing something?

✨ new CLI with click
✨ no more dicts, now data is handled with pydantic
✨ better output templating
✨ new commands: fav, file and search
♻️ Tidal api redesigned
✅ new tests

**Reply by oskvr37** [reply]
1. The updated CLI introduces several commands for an enhanced user experience.
```bash
$ tiddl
Usage: tiddl [OPTIONS] COMMAND [ARGS]...

  TIDDL - Download Tidal tracks ♫

Options:
  -v, --verbose  Show debug logs
  --help         Show this message and exit.

Commands:
  auth    Manage Tidal token
  config  Print path to the configuration file
  fav     Get your Tidal favorites
  file    Parse txt or JSON file with URLs
  search  Search on Tidal
  url     Get Tidal URL.
```
2. Template formatting
Output templating has been upgraded, making it more flexible and customizable. Check the [template formatting documentation](https://github.com/oskvr37/tiddl/wiki/Template-formatting) for details.

3. Codebase overhaul
Pydantic is now used for data handling instead of dictionaries.
The Tidal API has been redesigned.
The CLI is now separate from the package API, which improves maintainability.
Almost 90% of the codebase has been rewritten, and 2.x is not backward-compatible with 1.x.

In short, this release improves both user and developer experiences, with better structure, features, and extensibility

Not every feature I envisioned has made it into 2.0. For example, the "Advanced Tidal search functionality" isn't fully realized yet. While the search command works in the CLI, it’s fairly basic at this stage and doesn’t include filters or advanced options.

However, I plan to make the search functionality much more robust in the GUI part of Tiddl.

**Comment by Darkloke** [comment]
Thanks for the clarification. Can i ask for your plans regarding this features? I mean, are they in high/low priorities?
- Metadata for tracks
- Tracking failed downloads
- Simultaneous downloads

**Reply by oskvr37** [reply]
> Metadata for tracks

This has been implemented since v1.4. The only missing part is [fetching track credits](https://github.com/oskvr37/tiddl/discussions/46#discussioncomment-11279245), which we’ve already discussed and is on my radar.

> Tracking failed downloads

This requires a clearer definition. Do we track:

- All tracks that were supposed to download and failed?
- Skipped tracks that were already downloaded?
- Tracks that were not streamReady?

Once we define the scope, I can prioritize this feature accordingly.

> Simultaneous downloads

This one will be high priority. With the new 2.0 structure, it’s much easier to develop and implement such features, so you can expect this soon.

**Reply by oskvr37** [reply]
Remember to feel free about adding [new issues](https://github.com/oskvr37/tiddl/issues) with feature requests :)
This will be easier for me to prioritize things and track them better.

**Reply by Darkloke** [reply]
>This has been implemented since v1.4. The only missing part is https://github.com/oskvr37/tiddl/discussions/46#discussioncomment-11279245, which we’ve already discussed and is on my radar.

Thanks. Its just one of the features i look forward to the most. But I didn't want to be too pushy either. :-D

>This requires a clearer definition. Do we track:
>All tracks that were supposed to download and failed?

One of the serious issues that bothered me too much with the yaronzz downloader - in case of failed downloads i had to copy console output and manually search for errors/failed tracks. Since the output was limited in case of large queue some part was truncated. 
So i suppose yes, this is the first type of tracks - that were in queue, but failed due to internet connection/ API/ other problems.

In our previous discussion thread i cited streamripper database as an example of such feature implementation, but its up to you how to be here.

> Skipped tracks that were already downloaded?

Regarding this i would like to ask - when we are skipping - are we skipping basing only on existing filename or by something else? Can we compare, for example, file checksum with the file from server? Could here be a situation when the tidal updates the files, but the album id isnt changed? For example tracks in new album dont have metadata, but after some time tidal add it there - files changed, but we dont redownload them, coz we already have files with such names on the disk.

>Tracks that were not streamReady?

TBH i am not sure what it is... Not available due to subscription/region reasons?

>Remember to feel free about adding [new issues](https://github.com/oskvr37/tiddl/issues) with feature requests :)
> This will be easier for me to prioritize things and track them better.

Thanks. Wasnt sure what better to ask here or post the issue. :)

**Reply by oskvr37** [reply]
Tracks are skipped when filename already exists, this is the checking function:
https://github.com/oskvr37/tiddl/blob/dc040ab1707b6547901dfaeaf0a021be31a5bb5f/tiddl/utils.py#L96
There is also --noskip flag that tells tiddl to rewrite already downloaded tracks (it will redownload track stream to file and update its metadata)

> Can we compare, for example, file checksum with the file from server? 

Tracks metadata and track stream data are separate data in Tidal.

> For example tracks in new album dont have metadata, but after some time tidal add it there - files changed, but we dont redownload them, coz we already have files with such names on the disk.

Do you mean Tidal actually miss metadata? I thought the metadata is always in place and is not changed after some time.

> TBH i am not sure what it is... Not available due to subscription/region reasons?

Exactly, not "streamReady" tracks are unavailable for reasons like subscription/region restrictions. However, there may be other reasons that Tidal limits streaming which we’re unaware of.
Since we can’t download such tracks anyway, I’m not sure if it makes sense to include them in the "failed tracks" list. Let me know your thoughts on this.

Anyway I can add the functionality to collect failed tracks to list and then output it to json or txt file.

Feel free to explore the Tiddl source code to better understand how everything works under the hood. I’ve tried to make the code as clean and understandable as possible👌

**Reply by Darkloke** [reply]
>Do you mean Tidal actually miss metadata? I thought the metadata is always in place and is not changed after some time.

I spent a lot of time working with albums metadata, so i specifically tracked and marked albums which contained detailed metadata (like exact composer/writers for tracks for example) and which - are not. After some time later i redownloaded those albums and was surprised to see that albums which previously didnt have metadata, now have it. So yes - from my personal experience i would say metadata changes with time and tidal update it. Also probably you could notice that tidal changes album id while "visually" there are no changes there - same amount of tracks, same cover, etc., but anyway something is changed and may be it was metadata update for this album.

But this was year+ ago, so may be i am missing something, so please take this with the grain of salt. I suppose here would be useful more feedback from tidal users.

And my question about skipping existing tracks was asked following the logic of downloading usual file from internet - if the file didnt changed on the server - skip it, but i suppose here it wont be so simple.

> Since we can’t download such tracks anyway, I’m not sure if it makes sense to include them in the "failed tracks" list. Let me know your thoughts on this.

An option would be nice i think, so a user would decide by himself how he wants to treat them - skip, track or note. May be download them as blank files with special extension, so they could be easily found then using file search.

Tbh i dont pay much attention to such files since if they cant be downloaded due to region restrictions, then you cant do much about this (unless you have several subscriptions to select from). +From my experience such files are mostly from compilations or "non-album" tracks (like bonus tracks for movie OSTs). If you are interested in the whole album - it either available, either not.

**Reply by oskvr37** [reply]
Alright I get your point.

> Also probably you could notice that tidal changes album id

I saw some songs appear on my playlists unavailable anymore but they can be still found (with new id), I dont think that Tidal changes the id, maybe it's just uploaders/authors reupload the albums with new metadata or new track data.

The only way to find out if track data changed is to redownload full track and set metadata again.
To find out if track has new id (was reuploaded on Tidal), we would need to store the id in the downloaded file metadata and then compare it with current fetched id of track.

I'm thinking about database with all resources ids fetched by api, this would be our cache of Tidal backend, but I don't know if this would be an overkill for such a simple cli app.

> Tbh i dont pay much attention to such files since if they cant be downloaded due to region restrictions

I guess some people use VPNs to access restricted songs anyway, so that's not an issue for them

**Comment by Darkloke** [comment]
You closed the issue [there ](https://github.com/oskvr37/tiddl/issues/73#issuecomment-2637825856), so i wasnt sure where to continue :)

>From Tidal API there is no:
label
barcode
catalog number

I will try to clarify these things among community. There is a good [site ](https://harmony.pulsewidth.org.uk/) which uses different digital stores APIs to return general metadata info about albums, but now as i see Tidal changed something in its API and it seems unavailable.

As i remember label was displayed below the album, but its unavailable for fetching through the API?

>And some of the tags will work only with FLAC files.

I suppose you are talking here about VORBIS_COMMENT block? While other tags could work only for apple/itunes origin formats?

**Reply by oskvr37** [reply]
> As i remember label was displayed below the album

It is but Tidal API calls it the copyright, should `copyright` tag be same as `label` tag?

> And some of the tags will work only with FLAC files.

I mean the tags will be added only to FLAC files, tiddl can download .m4a and .flac, but these new tags that will be added only to .flac.
The python audio metadata module `mutagen` is unintuitive and in my opinion it's hard to work with - some of tags can be added to .m4a, some can be added to .flac but can't be added to .m4a, I wish I had time to write some new easy audio metadata module

> Tidal changed something in its API

Yeah they have API versioning but they change things in the way that the API suddenly stops working 🤷

[Let's continue in new discussion](https://github.com/oskvr37/tiddl/discussions/79)

**Comment by Darkloke** [comment]
So, multithreaded downloads now here as i understand? :)

**Reply by oskvr37** [reply]
Yes, multithreaded downloads are now supported in `v2.2.0`, and you can control the number of concurrent downloads using the -t option. For example `tiddl url <url> download -t 4`, 4 is the default count of threads.

**Reply by Darkloke** [reply]
Fantastic work, simply fantastic!

**Comment by Darkloke** [comment]
Hm, i may be missing something - is it possible to create a download template by adding {id} and {quality} to the {album}? Something like "Album 123456789 Master"? There are often albums with the same names, but different ids and quality, so this template would separate them when downloading large queues like discographies.

**Reply by oskvr37** [reply]
[Template formatting on wiki](https://github.com/oskvr37/tiddl/wiki/Template-formatting)

Currently no... but something new came to my mind, a new file templating system.
In the current function, I had to manually code the template keys.

https://github.com/oskvr37/tiddl/blob/56968be9a222f813db035c61397b2c28e448ef8a/tiddl/utils.py#L70-L90

But python `str.format` function can access object properties like this:
```py
"{album.title} {album.id} {album.quality}/{track.artist.name}/{track.title} {track.id}"
    .format(album=SomeAlbum(), track=SomeTrack())
```

With this approach, users will be able to access every attribute of resources and they won't be limited to hard coded values.
But there will be also aliases for some values for better user experience.

I will create feature request for this

**Reply by Darkloke** [reply]
Hey, any news @oskvr37? :)

**Reply by oskvr37** [reply]
Hey, currently tiddl development is stalled, we are waiting for any stable version of Tidal API.
Additionally I have started working abroad so actually I don't have enough time to write code...
But I will find free time if Tidal would release stable API finally :)
